homeassistant/components/melnor/__init__.py

```diff
@@ -0,0 +1,48 @@
+"""The melnor integration."""
+
+
+from melnor_bluetooth.device import Device
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_MAC, Platform
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator
+
+from .const import DOMAIN
+from .models import MelnorData, MelnorDataUpdateCoordinator
+
+PLATFORMS: list[Platform] = [Platform.SWITCH]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up melnor from a config entry."""
+    hass.data.setdefault(DOMAIN, {}).setdefault(entry.entry_id, {})
+
+    # Create the device and connect immediately so we can pull down
+    # required attributes before building out our entities
+    device = Device(mac=entry.data[CONF_MAC])
+
+    await device.connect()
```

Comment:
This has proven flaky in the real world in my dogfooding on my production HA box. I'm going to move this to the config flow so we pull the required attributes _once_ and store them in the config entry itself negating the need to do this heavier operation on every boot.

This will have an additional benefit of telling me in the config flow if the device is a valid and connectable melnor device right up front before passing to back to core to create the config entry.

Reply:
I refactored this code and now use `ConfigEntryNotReady` if for whatever reason we fail to establish an initial connection to the device. It ends up feeling really clean because a) we guarantee the device is connected and contains the required attributes before we init the entities and b) the backoff-retry is really nice. win win.